packages/framework/immutable-merge/README.md

```diff
@@ -3,82 +3,183 @@
-| `depth?: number`                          | Depth of recursion. Positive numbers will recurse that number of times, 0 will not recurse, a negative number will recurse infinitely. Unspecified is treated as 0.                                                                                                                                                                                |
-| `processSingles?: boolean`                | If true this will run through the tree even if there is only one viable branch. This is used to allow handlers to optionally update branches of the tree. This allows the routine to be used for things like processing all style objects in a complex tree, optionally changing them based on some logic, and returning a minimally updated tree. |
-| `recurse?: { [key] : boolean | handler }` | This allows overriding the normal handling based on the name of encountered keys. A value of true means that the routine will always recurse if that value is encountered. Specifying a handler function will cause that handler to be run when that key is encountered.                                                                           |
+## Deep Merge via `immutableMerge`
 
-## IMergeRecursionHandler
+For standard deep merging, this package provides the `immutableMerge` function. The signature is as follows:
 
-This is the signature for a handler function that can handle a named branch of the tree.
+```ts
+export function immutableMerge<T extends object>(...objs: (T | undefined)[]): T | undefined;
+```
 
-    export type IMergeRecursionHandler = (
-      key: string,
-      options: IMergeOptions,
-      ...objs: (object | undefined)[]) => object | undefined;
+This takes one or more objects of type `T` and deep merges them. If objects are undefined or null in some manner they will be ignored. Merging via this routine (and all routines in the package) typically follow the semantics of `Object.assign`, with a few extra behaviors.
 
-The array of objects or undefined values (internally treated as anything falsy) should be processed, returning a single object or undefined.
+- all values at a given level will overwrite, with the last writer winning
+- if a key does not exist for an object it is ignored
+- if a key does exist, even if it is `undefined` it will replace the previous value
+- only non-array objects will recurse, arrays will be replaced rather than appended
+- keys which exist and have a value of `undefined` will be deleted
 
-The key and options parameters are provided as conveniences in the case that a single handler needs to differentiate different branches or know how deep it is in the tree. In many cases these can be ignored.
+The peculiar pattern of deleting keys which end up as undefined is the only way to delete a key without violating the core principles. An example of key deletion might look like:
```

Comment:
If this is a common enough scenario, do we just want to add it as option in `MergeOptions`. That is allowing specification of keys to delete or not include in the final resultant object.

Reply:
MergeOptions is really only pattern matching via index signatures.  I could add a special name, maybe to change behavior for undefined but that is getting weird.

Reply:
Yeah I agree, it would be weird to do without making MergeOptions a heavier object with a full schema. The pattern is probably fine.